lib/api/graphql/link/custom_refresh_link.dart

```diff
@@ -0,0 +1,131 @@
+import 'package:pedantic/pedantic.dart';
+
+typedef ShouldRefresh = bool Function(Response);
+
+typedef RefreshToken<T> = Future<T> Function(T);
+
+/// {@template fresh_link}
+/// A GraphQL Link which handles manages an authentication token automatically.
+///
+/// A constructor that returns a Fresh interceptor that uses the
+/// [OAuth2Token] token, the standard token class and define the`
+/// tokenHeader as 'authorization': '${token.tokenType} ${token.accessToken}'
+///
+/// ```dart
+/// final freshLink = FreshLink(
+///   tokenStorage: InMemoryTokenStorage(),
+///   doRefreshing: (token) {
+///     // Perform refresh and return new token
+///   },
+/// );
+/// final graphQLClient = GraphQLClient(
+///   cache: InMemoryCache(),
+///   link: Link.from([freshLink, HttpLink(uri: 'https://my.graphql.api')]),
+/// );
+/// ```
+/// {@endtemplate}
+class CustomRefreshLink<T> extends Link with FreshMixin<T> {
+  /// {@macro fresh_link}
+  CustomRefreshLink({
+    @required TokenStorage<T> tokenStorage,
+    @required RefreshToken<T> refreshToken,
+    @required ShouldRefresh shouldRefresh,
+    TokenHeaderBuilder<T> tokenHeader,
+  })  : assert(tokenStorage != null),
+        assert(refreshToken != null),
+        assert(shouldRefresh != null),
+        _refreshToken = refreshToken,
+        _tokenHeader = tokenHeader,
+        _shouldRefresh = shouldRefresh {
+    this.tokenStorage = tokenStorage;
+  }
+
+  ///{@template fresh_link}
+  ///A GraphQL Link which handles manages an authentication token automatically.
+  ///
+  /// ```dart
+  /// final freshLink = FreshLink.oAuth2(
+  ///   tokenStorage: InMemoryTokenStorage<OAuth2Token>(),
+  ///   refreshToken: (token, client) {
+  ///     // Perform refresh and return new token
+  ///   },
+  /// );
+  /// final graphQLClient = GraphQLClient(
+  ///   cache: InMemoryCache(),
+  ///   link: Link.from([freshLink, HttpLink(uri: 'https://my.graphql.api')]),
+  /// );
+  /// ```
+  /// {@endtemplate}
+  static CustomRefreshLink<OAuth2Token> oAuth2({
+    @required TokenStorage<OAuth2Token> tokenStorage,
+    @required RefreshToken<OAuth2Token> doRefreshing,
+    @required ShouldRefresh shouldRefresh,
+    TokenHeaderBuilder<OAuth2Token> tokenHeader,
+  }) {
+    return CustomRefreshLink<OAuth2Token>(
+      refreshToken: doRefreshing,
+      tokenStorage: tokenStorage,
+      shouldRefresh: shouldRefresh,
+      tokenHeader: tokenHeader ??
+          (token) {
+            return {
+              'authorization': '${token.tokenType} ${token.accessToken}',
+            };
+          },
+    );
+  }
+
+  final RefreshToken<T> _refreshToken;
+  final TokenHeaderBuilder<T> _tokenHeader;
+  final ShouldRefresh _shouldRefresh;
+
+  @override
+  Stream<Response> request(
+    Request request, [
+    NextLink forward,
+  ]) async* {
+    final currentToken = await token;
+    final tokenHeaders = currentToken != null && _tokenHeader != null
+        ? _tokenHeader(currentToken)
+        : const <String, String>{};
```

Comment:
actually, I feel like I can skip the AuthLink and just pass the logic to here for the token header appending 🤔  let me recheck it a bit

Reply:
I was right, I will simplify the implementation and making a `Header Interceptor` + `Authenticator` on failure into this single Link, please help me review again on the next push, @manh-t @markgravity 🙇

Reply:
@sleepylee it's good to go 🙏